src/core/transport/ssu_packet.h

```diff
@@ -769,175 +771,87 @@ class SSUPacketParser {
-
-/// @brief Writes RelayRequest message
-/// @param buf Reference to pointer to buffer to write into
-/// @param packet RelayRequest packet to write
-void WriteRelayRequest(
-    std::uint8_t*& buf,
-    SSURelayRequestPacket* packet);
-
-/// @brief Writes RelayResponse message
-/// @param buf Reference to pointer to buffer to write into
-/// @param packet RelayResponse packet to write
-void WriteRelayResponse(
-    std::uint8_t*& buf,
-    SSURelayResponsePacket* packet);
-
-/// @brief Writes RelayIntro message
-/// @param buf Reference to pointer to buffer to write into
-/// @param packet RelayIntro packet to write
-void WriteRelayIntro(
-    std::uint8_t*& buf,
-    SSURelayIntroPacket* packet);
-
-/// @brief Writes Data message
-/// @param buf Reference to pointer to buffer to write into
-/// @param packet Data packet to write
-void WriteData(
-    std::uint8_t*& buf,
-    SSUDataPacket* packet);
-
-/// @brief Writes PeerTest message
-/// @param buf Reference to pointer to buffer to write into
-/// @param packet PeerTest packet to write
-void WritePeerTest(
-    std::uint8_t*& buf,
-    SSUPeerTestPacket* packet);
-
-/// @brief Writes SessionDestroyed message
-/// @param buf Reference to pointer to buffer to write into
-/// @param packet SessionDestroyed packet to write
-void WriteSessionDestroyed(
-    std::uint8_t*& buf,
-    SSUSessionDestroyedPacket* packet);
-
-/// @brief Writes SSU packet for SSU session
-/// @param buf Reference to pointer to buffer to write into
-/// @param packet SSU packet to write
-/// @note packet is one of any payload types
-void WritePacket(
-    std::uint8_t*& buf,
-    SSUPacket* packet);
-}  // namespace SSUPacketBuilder
+/// @class SSUPacketBuilder
+/// @brief Constitutes SSU packet building
+class SSUPacketBuilder : private i2p::util::OutputByteStream {
+  using i2p::util::OutputByteStream::WriteData; // Overload resolution
+ public:
+  using i2p::util::OutputByteStream::GetPosition; // Make public 
+
+  SSUPacketBuilder() = default;
+
+  /// @brief Constructs packet builder with a given buffer
+  /// @param data Pointer to the first byte of the buffer 
+  /// @param len Length of the buffer 
+  SSUPacketBuilder(
+      std::uint8_t* data,
+      std::size_t len);
+
+  /// @brief Calculates padding size needed for message
+  /// @details All messages contain 0 or more bytes of padding.
+  ///   Each message must be padded to a 16 byte boundary,
+  ///   as required by the AES256 encryption layer
+  /// @param size Size of message
+  static std::size_t GetPaddingSize(
+      std::size_t size);
+
+  /// @brief Gets padded size of message
+  /// @param size Size of message
+  static std::size_t GetPaddedSize(
+      std::size_t size);
+
+  /// @brief Writes an SSU header into a data buffer.
+  /// @pre The data buffer must be sufficiently large.
+  /// @param data Reference to pointer to data
+  /// @param header Pointer to SSU header
+  void WriteHeader(SSUHeader* header);
```

Comment:
Passing a `unique_ptr` implies that you move ownership, which you don't want to do in this situation. I pass a raw pointer here because I don't need any memory management, and because I want to allow that `header` can be a reference to stack memory (in which case any `delete` would be inappropriate).

What we should consider, though, is to pass `header` as a pointer-to-const.